File: cl/scala_cl/cl2.scala
import scala.annotation.tailrec

@main
def main(): Unit = {
  println(fib(5))
  println(fibTail(5))
  println(initSegment(List(1,2,3), List(1,2,3)))
  println(initSegment(List(1,2,3), List(1,3,3)))
  println(replaceNth(List(1,2,3,4,5), 2, 5))

}
//dziala
def fib(n:Int): Int = n match {
  case 0 => 0
  case 1 => 1
  case _ => fib (n-1) + fib (n-2)
}
def fibTail(n:Int): Int = {
  @tailrec
  def fibTailHelper(n:Int, curr: Int, prev: Int): Int = {
    n match {
      case 0 => prev
      case 1 => curr
      case _ => fibTailHelper(n-1,curr+prev,curr)
    }
  }
  fibTailHelper(n,1,0)
}
@tailrec
def initSegment[A](xs:List[A], ys:List[A]): Boolean = {
  (xs, ys) match {
    case (Nil,_) => true
    case (_,Nil) => false
    case _ => if xs.head == ys.head then initSegment(xs.tail,ys.tail)
    else false
  }
}

def replaceNth[A](xs:List[A], n:Int, x:A): List[A] = {
  (xs, n) match {
    case (Nil, _) => Nil
    case (_ :: tail, 0) => x :: tail
    case (head :: tail, _) => head :: replaceNth(tail, n - 1, x)
  }
}

File: cl/ocaml_cl/cl2.ml
let rec fib n = 
  match n with
  | 0 -> 0
  | 1 -> 1
  | _ -> fib (n-1) + fib (n-2)
;;

let rec fibTailHelper n curr prev =
  match n with
  | 0 -> prev
  | 1 -> curr
  | _ -> fibTailHelper (n-1) (curr+prev) curr
;;

let rec fibTail n = 
  let rec fibTailHelper n curr prev =
    match n with
    | 0 -> prev
    | 1 -> curr
    | _ -> fibTailHelper (n-1) (curr+prev) curr
  in fibTailHelper n 1 0
;;

let rec initSegment xs ys =
  match xs,ys with 
  | ([],_) -> true
  | (_,[]) -> false
  | (h1::tl1,h2::tl2) -> if h1 = h2 then initSegment tl1 tl2 else false;;


let rec replaceNth xs n x = 
  match (xs,n) with 
    | ([],_) -> [x]
    | (_::tail, 0) -> x::tail
    | (head::tail, _) -> head::replaceNth tail (n-1) x
  ;;

  let resultFib = fib 5;;
let resultFibTail = fibTail 5;;
let resultIS1 = initSegment [] [1; 2; 3; 4] ;;
let resultIS2 = initSegment [1;2;3] [1;2;3];;
let list = [1;2;3;4;5];;
let resultRN = replaceNth list 2 5;;